src/state/slices/opportunitiesSlice/resolvers/idle.ts

```diff
@@ -0,0 +1,100 @@
+import type { AssetId } from '@shapeshiftoss/caip'
+import { fromAssetId, toAssetId } from '@shapeshiftoss/caip'
+import { DefiProvider, DefiType } from 'features/defi/contexts/DefiManagerProvider/DefiCommon'
+import { getIdleInvestor } from 'features/defi/contexts/IdleProvider/idleInvestorSingleton'
+import { logger } from 'lib/logger'
+import { selectAssetById } from 'state/slices/selectors'
+
+import type {
+  GetOpportunityIdsOutput,
+  GetOpportunityMetadataOutput,
+  OpportunitiesState,
+  OpportunityDefiType,
+  StakingId,
+} from '../types'
+import type { ReduxApi } from './types'
+
+const moduleLogger = logger.child({ namespace: ['opportunities', 'resolvers', 'idle'] })
+
+export const idleStakingOpportunitiesMetadataResolver = async ({
+  opportunityType,
+  reduxApi,
+}: {
+  opportunityType: OpportunityDefiType
+  reduxApi: ReduxApi
+}): Promise<{ data: GetOpportunityMetadataOutput }> => {
+  const idleInvestor = getIdleInvestor()
+  const opportunities = await idleInvestor.findAll()
+
+  const { getState } = reduxApi
+  const state: any = getState() // ReduxState causes circular dependency
+
+  const byId = await opportunities.reduce(
```

Comment:
After pondering this some more, I think reducing makes this more complicated than it needs to be.

See a suggested change with a for..of mutating an object:

```ts
const metaDataById: OpportunitiesState[DefiType.Staking]['byId'] = {}
  for (const opportunity of opportunities) {
    const assetId = toAssetId({
      assetNamespace: 'erc20',
      assetReference: opportunity.id,
      chainId: fromAssetId(opportunity.feeAsset.assetId).chainId,
    })
    const asset = selectAssetById(state, assetId)

    // Asset doesn't exist in portfolio, meaning this asset is bogus, e.g these two
    // https://etherscan.io/address/0xa0154a44c1c45bd007743fa622fd0da4f6d67d57
    // https://etherscan.io/address/0x5f45a578491a23ac5aee218e2d405347a0fafa8e
    if (!asset) continue

    const rewardAssetIds = (await opportunity.getRewardAssetIds().catch(error => {
      moduleLogger.debug(
        { fn: 'idleStakingOpportunitiesMetadataResolver', error },
        'Error fetching Idle opportunities metadata',
      )
    })) as [AssetId, AssetId?, AssetId?] | undefined

    metaDataById[assetId as StakingId] = {
      apy: opportunity.apy.toString(),
      assetId,
      provider: DefiProvider.Idle,
      tvl: opportunity.tvl.balance.toString(),
      type: DefiType.Staking,
      underlyingAssetId: assetId,
      underlyingAssetIds: [opportunity.underlyingAsset.assetId],
      ...{ rewardAssetIds },
      // Idle opportunities wrap a single yield-bearing asset, so the ratio will always be 1
      underlyingAssetRatios: ['1'],
      name: asset.name,
    }
  }
```

Reply:
While I'm a fan of the async promise reducer pattern, I have to agree it can get confusing to read!
Coupled with the fact that `for ... of` is a construct that has first-class async iterators support and is going to be more performant, mutating is a tradeoff we can live with 😄 

b7aea3742c3c0114546a54676a47cbb303aed451